frontend/catalyst/api_extensions/function_maps.py

```diff
@@ -234,74 +185,166 @@ def loop_fn(i, batched_result_list):
-    indicating where the mapping should apply.
+        Returns:
+            List: A list of indices representing the locations where the mapping should be applied.
+        """
 
-    Args:
-        axes_flat (List): Flattened list of in-axes or out-axes including `None` elements.
+        return [i for i, d in enumerate(axes_flat) if d is not None]
 
-    Returns:
-        List: A list of indices representing the locations where the mapping should be applied.
-    """
+    def _get_batch_size(self, args_flat, axes_flat, axis_size):
+        """Get the batch size based on the provided arguments and axes specifiers, or the manually
+        specified batch size by the user request. The batch size must be the same for all arguments.
+
+        Args:
+            args_flat (List): Flatten list of arguments.
+            axes_flat (List): Flatten list of `in_axes` or `our_axes` including `None` elements.
+            axis_size (Optional[int]): Optional default batch size.
+
+        Returns:
+            int: Returns the batch size used as the upper bound of the QJIT-compatible for loop
+                in the computation of vmap.
+
+        Raises:
+            ValueError: The batch size must be the same for all arguments.
+            ValueError: The default batch is expected to be None, or less than or equal
+            to the computed batch size.
+        """
+
+        batch_sizes = []
+        for arg, d in zip(args_flat, axes_flat):
+            shape = np.shape(arg) if arg.shape else (1,)
+            if d is not None and len(shape) > d:
+                batch_sizes.append(shape[d])
+
+        if any(size != batch_sizes[0] for size in batch_sizes[1:]):
+            raise ValueError(
+                "Invalid batch sizes; expected the batch size to be the same for all arguments, "
+                f"but got batch_sizes={batch_sizes} from args_flat={args_flat}"
+            )
+
+        batch_size = batch_sizes[0] if batch_sizes else 0
 
-    return [i for i, d in enumerate(axes_flat) if d is not None]
+        if axis_size is not None:
+            if axis_size <= batch_size:
+                batch_size = axis_size
+            else:
+                raise ValueError(
+                    "Invalid 'axis_size'; the default batch is expected to be None, "
+                    "or less than or equal to the computed batch size, but got "
+                    f"axis_size={axis_size} > batch_size={batch_size}"
+                )
+
+        if not batch_size:
+            raise ValueError(
+                f"Invalid batch size; it must be a non-zero integer, but got {batch_size}."
+            )
 
+        return batch_size
 
-def _get_batch_size(args_flat, axes_flat, axis_size):
-    """Get the batch size based on the provided arguments and axes specifiers, or the manually
-       specified batch size by the user request. The batch size must be the same for all arguments.
+
+## API ##
+def vmap(
```

Comment:
Just a note if we to stick to the convention in the api_extensions, the API functions are at the top of the files.

Reply:
Done